liquidator/liquidator.js

```diff
@@ -158,12 +168,50 @@ class Liquidator {
       // Note: query the time again during each iteration to ensure the deadline is set reasonably.
       const currentBlockTime = this.empClient.getLastUpdateTime();
 
+      // Calculate the amount of tokens we will attempt to liquidate.
+      let tokensToLiquidate;
+
+      // If the user specifies `maxTokensToLiquidateWei`, then we must make sure that it follows an important constraint:
+      // we cannot bring the position down below the `minSponsorPosition`.
+      if (maxTokensToLiquidateWei) {
+        // First, we check if `maxTokensToLiquidateWei > tokensOutstanding`, if so then we'll liquidate the full position.
+        if (this.toBN(maxTokensToLiquidateWei).gte(this.toBN(position.numTokens))) {
+          tokensToLiquidate = this.toBN(position.numTokens);
+        } else {
+          // If we're not liquidating the full position, then we cannot liquidate the position below the `minSponsorTokens` constraint.
+          // `positionTokensAboveMinimum` is the maximum amount of tokens any liquidator can liquidate while taking `minSponsorTokens` into account.
+          const positionTokensAboveMinimum = this.toBN(position.numTokens).sub(this.toBN(this.empMinSponsorSize));
+
+          // Finally, we cannot liquidate more than `maxTokensToLiquidate`.
+          tokensToLiquidate = this.BN.min(positionTokensAboveMinimum, this.toBN(maxTokensToLiquidateWei));
+        }
+      } else {
+        // If `maxTokensToLiquidateWei` is not specified, then we will attempt to liquidate the full position.
+        tokensToLiquidate = this.toBN(position.numTokens);
+      }
+
+      // If `tokensToLiquidate` is 0, then skip this liquidation.
+      if (tokensToLiquidate.isZero()) {
```

Comment:
We have to check for this because its possible to call `createLiquidation(...tokensOutstanding=0...)

Reply:
so this will only happen if the wallet has been drained to zero. that's good. however should this log message also not be fired also if the `positionTokensAboveMinimum> tokensToLiquidate`? would this not also catch this zero case?

Reply:
>so this will only happen if the wallet has been drained to zero.

Important distinction here: no. This will only happen if the bot intends to liquidate zero tokens. For example imagine that the bot wants to liquidate 4 tokens, but the minimum sponsor size is 5 tokens, and the current position size is 5 tokens. Now, look at the if-statement branching in the code: 
1. `maxTokensToLiquidateWei (4) >= position.numTokens (5)` is FALSE 
2. `positionTokensAboveMinimum = position.numTokens (5) - empMinSponsorSize (5) = 0`
3. `tokensToLiquidate = MIN(positionTokensAboveMinimum (0), maxTokensToLiquidateWei (5)) = 0`

So in this scenario, the bot would only be allowed to liquidate 0 tokens, which is possible, but we need to catch. This is exactly the scenario I accidentally encountered while testing and realized I had to catch.

>however should this log message also not be fired also if the positionTokensAboveMinimum > tokensToLiquidate

If `positionTokensAboveMinimum > tokensToLiquidate`, then the bot would just try to liquidate `tokensToLiquidate`. The bot will liquidate the minimum of those two values. For example, if the position size is 8, the bot wants to liquidate 2 tokens, and the minimum sponsor size is 5, then the `positionTokensAboveMinimum = 8-5 = 3`, but the bot only has permission to liquidate 2 tokens. So it will liquidate 2 tokens to bring the position size to 6.

Reply:
ok great, thanks for clearing that up! complex logic going on here. This branching logic is the most complex part of these bots so it's important it's well commented and tested so that's for doing such a good job on that.